meson.build

```diff
@@ -243,6 +243,118 @@ if os_family_name == 'BSD'
+
+if zlib_ng_gets_native
+    extra_flags += '-march=native'
+    simd_summary = 'native'
+    message('Enabling native SIMD optimizations')
+
+elif zlib_ng_wants_native
+    # Wanted native SIMD but couldn't get it, so
+    # lets test and enable some basic ones for them.
+    #
+    # NEON
+    # ~~~~~
+    neon_cflag = '-mfpu=neon'
+    if target_machine.cpu_family() == 'aarch64'
+        # aarch64 always supports NEON:
+        # ref: https://developer.arm.com/documentation/den0024/a/AArch64-Floating-point-and-NEON
+        zlib_ng_options += ['neon']
+        extra_flags += neon_cflag
+        simd_summary = 'NEON'
+        message('Host supports ARM NEON instruction set')
+
+    elif host_machine.cpu_family() == 'arm'
+        neon_test_code = '''
+            #include <arm_neon.h>
+            int main() {
+            uint8x16_t a = vdupq_n_u8(0);
+            uint8x16_t b = vdupq_n_u8(0);
+            uint8x16_t result = vaddq_u8(a, b);
+            return 0;
+            }
+        '''
+        neon_test = cc.run(
+            neon_test_code,
+            args: neon_cflag,
+            name: 'ARM NEON instruction set test',
+        )
+        if (neon_test.compiled() and neon_test.returncode() == 0)
+            zlib_ng_options += ['neon']
+            extra_flags += neon_cflag
+            simd_summary = 'NEON'
+            message('Enabling the ARM NEON instruction set')
+        endif
+
+    elif target_machine.cpu_family().startswith('x86')
+        # SSE2
+        # ~~~~
+        sse2_test_code = '''
+            #include <emmintrin.h>
+            int main() {
+            __m128i a = _mm_setzero_si128();
+            __m128i b = _mm_setzero_si128();
+            __m128i result = _mm_add_epi32(a, b);
+            return 0;
+            }
+        '''
+        sse2_cflag = '-msse2'
+        sse2_test = cc.run(
+            sse2_test_code,
+            args: sse2_cflag,
+            name: 'SSE2 instruction set test',
+        )
+        if (sse2_test.compiled() and sse2_test.returncode() == 0)
+            zlib_ng_options += ['sse2']
+            extra_flags += sse2_cflag
+            simd_summary = 'SSE2'
+            message('Enabling the SSE2 instruction set')
+        endif
+
+        # SSSE3
+        # ~~~~~
+        ssse3_test_code = '''
+            #include <tmmintrin.h>
+            int main() {
+            __m128i a = _mm_setzero_si128();
+            __m128i b = _mm_setzero_si128();
+            __m128i result = _mm_hadd_epi16(a, b);
+            return 0;
+            }
+        '''
+        ssse3_cflag = '-mssse3'
+        ssse3_test = cc.run(
+            ssse3_test_code,
+            args: ssse3_cflag,
+            name: 'SSSE3 instruction set test',
+        )
+        if (ssse3_test.compiled() and ssse3_test.returncode() == 0)
+            zlib_ng_options += ['ssse3']
+            extra_flags += ssse3_cflag
+            simd_summary = 'SSE2 and SSSE3'
+            message('Enabling the SSSE3 instruction set')
```

Comment:
Huh, this all seems rather complicated; isn't there a simpler way?

Reply:
Unfortunately not; the compiler will happily generate the instructions but execution on the target is the only way test them. These are tiny zero-dependency builds, so they fly by during the setup phase (and aren't tested again for all subsequent builds).

Reply:
Ok, I'm far from being a build automation expert, so I'll trust you on this 😄 Seemed a bit like a maintenance burden (perhaps) but if you're saying it's necessary, I'm fine with it.